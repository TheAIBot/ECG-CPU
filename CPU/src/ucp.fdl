ipblock circularArray(in address : ns(32);
                      in wr,rd : ns(1);
                      in idata : tc(32);
	              out odata : tc(32))
{
	iptype "ram";
	ipparm "size=32";
	ipparm "wl=32";
}

dp mwiFilter	(in datain : tc(32);
		 in datainrdy : ns(1);
		 out dataout : tc(32);
		 out dataoutrdy : ns(1))
{
	//The registers used in calculations:
	reg difference : tc(32); //Reg
	reg formerValue : tc(32); //R6 
	reg arrayIndex : ns(5); //R1
	reg sum : tc(32); //R3
	
	reg hasDataIn : ns (1); //Used by the final state machine	

	sig updatedSum : tc(32); //As the name says
	
	//For the connection to the memory block/circular array
	sig circArrayOut : tc(32);
	sig indexPP : ns(5);	
	sig writeCircularArray : ns(1);
	sig readCircularArray : ns(1);
	use circularArray(indexPP,writeCircularArray,readCircularArray, datain, circArrayOut);
	
	//Though it stalls, it calculates the values for cycle 1.
	sfg stall{		
		//Calculates the difference from the former value on the place in teh circular array, and the new input. 
		difference = datain - formerValue; 

		//Goes to the next cycle if datainrdy = 1 ((*) same as the other UCP versions)
		hasDataIn = datainrdy; 
		
		//Writes datain to the circular array at indexPP=arrayIndex.
		writeCircularArray = 1;
		readCircularArray = 0;
		indexPP = arrayIndex;

		//No output is ready.
		dataoutrdy = 0;
		dataout = 0;
	}
	sfg addSumShift{ //Calculates the result of the UCP and gives it as an output.
		updatedSum = sum + difference; //Says it on the tin
		sum = updatedSum;

		//Bitshifts the updated sum by 5, and gives it as an output.
		dataout = (tc(32))updatedSum[5:31]; 
		dataoutrdy = 1;

		//Reads the next element in the circular array, and updates the current index.
		indexPP = arrayIndex + 1;
		arrayIndex = indexPP;		
		writeCircularArray = 0;
		readCircularArray = 1;
	}
	sfg setLoadedValue{ //Gets ready for a new input.
		formerValue = circArrayOut; //The output with the delay of one cycle, is now ready.
		writeCircularArray = 0;
		readCircularArray = 0;
		indexPP = arrayIndex;
		hasDataIn = 0;		
		dataoutrdy = 0;
		dataout = 0;
	}
	always
	{	
		//$display($dec, "Cycle = ", $cycle);
		//$display($dec, "In UCP, datain = ", datain, ", datainrdy = ", datainrdy, ", dataout = ", dataout, ", dataoutrdy = ", dataoutrdy);
		//$display($dec, "Also in UCP, hasDataIn = ", hasDataIn, ", difference = ", difference, ", sum = ", sum, ", arrayindex = ", arrayIndex);
	}
}

fsm ucp(mwiFilter){
	initial sStall;
	state sCycle2;
	
	@sStall  	if(hasDataIn) 	then		(addSumShift) 		-> sCycle2;
		 	else				(stall)        		-> sStall;
	@sCycle2 					(setLoadedValue)  	-> sStall; //And ready for the next round of inputs.
}














