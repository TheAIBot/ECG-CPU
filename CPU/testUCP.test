#!/bin/bash
name="testUCP"
cpp -P $name".fdl" | fdlsim 105 > result.txt
java -jar testVerifier.jar $name".expected" result.txt
if [ $? = "0" ]; then
    echo "Success $name"
else
	echo "Failed $name"
fi
