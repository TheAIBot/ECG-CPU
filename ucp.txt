

//CODE

LOAD R6 R1 CACHE
SUB R6 R5 R6

STORE R5 R1 CACHE
ADD R3 R3 R6
SRI R7 R3 5
ADDI R1 R1 1
AND R1 R1 R2


//COMMAND

LOAD R1 R2 UCP STALL = COP R1 R2 STALL

