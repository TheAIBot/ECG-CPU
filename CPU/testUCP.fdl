#include "src/ucp.fdl"

dp testUCP  (out datain : tc(32);
             out datainrdy : ns(1);
             in dataout : tc(32);
             in dataoutrdy : ns(1)){
    //1
    sfg dataset1In{datain = 0; datainrdy = 1;}
    sfg dataset1Out{$display("For dataset1, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0, giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset1Wait{datain = 0; datainrdy = 0;}   
    //2
    sfg dataset2In{datain = 0; datainrdy = 1;}
    sfg dataset2Out{$display("For dataset2, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0, giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset2Wait{datain = 0; datainrdy = 0;}    
    //3
    sfg dataset3In{datain = 0; datainrdy = 1;}
    sfg dataset3Out{$display("For dataset3, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0, giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset3Wait{datain = 0; datainrdy = 0;}    
    //4
    sfg dataset4In{datain = 0; datainrdy = 1;}
    sfg dataset4Out{$display("For dataset4, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0, giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset4Wait{datain = 0; datainrdy = 0;}    
    //5
    sfg dataset5In{datain = 1; datainrdy = 1;}
    sfg dataset5Out{$display("For dataset5, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0, giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset5Wait{datain = 0; datainrdy = 0;}    
    //6
    sfg dataset6In{datain = 4; datainrdy = 1;}
    sfg dataset6Out{$display("For dataset6, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0, giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset6Wait{datain = 0; datainrdy = 0;}    
    //7
    sfg dataset7In{datain = 9; datainrdy = 1;}
    sfg dataset7Out{$display("For dataset7, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0, giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset7Wait{datain = 0; datainrdy = 0;}    
    //8
    sfg dataset8In{datain = 16; datainrdy = 1;}
    sfg dataset8Out{$display("For dataset8, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 0 giving: ", dataout == 0);
                    datain = 0; datainrdy = 0;}
    sfg dataset8Wait{datain = 0; datainrdy = 0;}    
    //9
    sfg dataset9In{datain = 25; datainrdy = 1;}
    sfg dataset9Out{$display("For dataset9, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 1, giving: ", dataout == 1);
                    datain = 0; datainrdy = 0;}
    sfg dataset9Wait{datain = 0; datainrdy = 0;}    
    //10
    sfg dataset10In{datain = 36; datainrdy = 1;}
    sfg dataset10Out{$display("For dataset10, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 2, giving: ", dataout == 2);
                    datain = 0; datainrdy = 0;}
    sfg dataset10Wait{datain = 0; datainrdy = 0;}    
    //11
    sfg dataset11In{datain = 49; datainrdy = 1;}
    sfg dataset11Out{$display("For dataset11, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 4, giving: ", dataout == 4);
                    datain = 0; datainrdy = 0;}
    sfg dataset11Wait{datain = 0; datainrdy = 0;}   
    //12
    sfg dataset12In{datain = 49; datainrdy = 1;}
    sfg dataset12Out{$display("For dataset12, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 5, giving: ", dataout == 5);
                    datain = 0; datainrdy = 0;}
    sfg dataset12Wait{datain = 0; datainrdy = 0;}   
    //13
    sfg dataset13In{datain = 49; datainrdy = 1;}
    sfg dataset13Out{$display("For dataset13, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 7, giving: ", dataout == 7);
                    datain = 0; datainrdy = 0;}
    sfg dataset13Wait{datain = 0; datainrdy = 0;}   
    //14
    sfg dataset14In{datain = 49; datainrdy = 1;}
    sfg dataset14Out{$display("For dataset14, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 8, giving: ", dataout == 8);
                    datain = 0; datainrdy = 0;}
    sfg dataset14Wait{datain = 0; datainrdy = 0;}   
    //15
    sfg dataset15In{datain = 49; datainrdy = 1;}
    sfg dataset15Out{$display("For dataset15, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 10, giving: ", dataout == 10);
                    datain = 0; datainrdy = 0;}
    sfg dataset15Wait{datain = 0; datainrdy = 0;}   
    //16
    sfg dataset16In{datain = 49; datainrdy = 1;}
    sfg dataset16Out{$display("For dataset16, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 12, giving: ", dataout == 12);
                    datain = 0; datainrdy = 0;}
    sfg dataset16Wait{datain = 0; datainrdy = 0;}   
    //17
    sfg dataset17In{datain = 25; datainrdy = 1;}
    sfg dataset17Out{$display("For dataset17, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 12, giving: ", dataout == 12);
                    datain = 0; datainrdy = 0;}
    sfg dataset17Wait{datain = 0; datainrdy = 0;}   
    //18
    sfg dataset18In{datain = 1; datainrdy = 1;}
    sfg dataset18Out{$display("For dataset18, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 12, giving: ", dataout == 12);
                    datain = 0; datainrdy = 0;}
    sfg dataset18Wait{datain = 0; datainrdy = 0;}   
    //19
    sfg dataset19In{datain = 25; datainrdy = 1;}
    sfg dataset19Out{$display("For dataset19, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 13, giving: ", dataout == 13);
                    datain = 0; datainrdy = 0;}
    sfg dataset19Wait{datain = 0; datainrdy = 0;}   
    //20
    sfg dataset20In{datain = 144; datainrdy = 1;}
    sfg dataset20Out{$display("For dataset20, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 18, giving: ", dataout == 18);
                    datain = 0; datainrdy = 0;}
    sfg dataset20Wait{datain = 0; datainrdy = 0;}   
    //21
    sfg dataset21In{datain = 400; datainrdy = 1;}
    sfg dataset21Out{$display("For dataset21, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 30, giving: ", dataout == 30);
                    datain = 0; datainrdy = 0;}
    sfg dataset21Wait{datain = 0; datainrdy = 0;}  
    //22
    sfg dataset22In{datain = 784; datainrdy = 1;}
    sfg dataset22Out{$display("For dataset22, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 55, giving: ", dataout == 55);
                    datain = 0; datainrdy = 0;}
    sfg dataset22Wait{datain = 0; datainrdy = 0;}  
    //23
    sfg dataset23In{datain = 1089; datainrdy = 1;}
    sfg dataset23Out{$display("For dataset23, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 89, giving: ", dataout == 89);
                    datain = 0; datainrdy = 0;}
    sfg dataset23Wait{datain = 0; datainrdy = 0;}  
    //24
    sfg dataset24In{datain = 1089; datainrdy = 1;}
    sfg dataset24Out{$display("For dataset24, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 123, giving: ", dataout == 123);
                    datain = 0; datainrdy = 0;}
    sfg dataset24Wait{datain = 0; datainrdy = 0;}  
    //25
    sfg dataset25In{datain = 841; datainrdy = 1;}
    sfg dataset25Out{$display("For dataset25, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 149, giving: ", dataout == 149);
                    datain = 0; datainrdy = 0;}
    sfg dataset25Wait{datain = 0; datainrdy = 0;}  
    //26
    sfg dataset26In{datain = 484; datainrdy = 1;}
    sfg dataset26Out{$display("For dataset26, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 164, giving: ", dataout == 164);
                    datain = 0; datainrdy = 0;}
    sfg dataset26Wait{datain = 0; datainrdy = 0;}  
    //27
    sfg dataset27In{datain = 196; datainrdy = 1;}
    sfg dataset27Out{$display("For dataset27, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 170, giving: ", dataout == 170);
                    datain = 0; datainrdy = 0;}
    sfg dataset27Wait{datain = 0; datainrdy = 0;}  
    //28
    sfg dataset28In{datain = 25; datainrdy = 1;}
    sfg dataset28Out{$display("For dataset28, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 171, giving: ", dataout == 171);
                    datain = 0; datainrdy = 0;}
    sfg dataset28Wait{datain = 0; datainrdy = 0;}  
    //29
    sfg dataset29In{datain = 1; datainrdy = 1;}
    sfg dataset29Out{$display("For dataset29, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 171, giving: ", dataout == 171);
                    datain = 0; datainrdy = 0;}
    sfg dataset29Wait{datain = 0; datainrdy = 0;}  
    //30
    sfg dataset30In{datain = 25; datainrdy = 1;}
    sfg dataset30Out{$display("For dataset30, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 172, giving: ", dataout == 172);
                    datain = 0; datainrdy = 0;}
    sfg dataset30Wait{datain = 0; datainrdy = 0;}  
    //31
    sfg dataset31In{datain = 64; datainrdy = 1;}
    sfg dataset31Out{$display("For dataset31, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 174, giving: ", dataout == 174);
                    datain = 0; datainrdy = 0;}
    sfg dataset31Wait{datain = 0; datainrdy = 0;}  
    //32
    sfg dataset32In{datain = 81; datainrdy = 1;}
    sfg dataset32Out{$display("For dataset32, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 176, giving: ", dataout == 176);
                    datain = 0; datainrdy = 0;}
    sfg dataset32Wait{datain = 0; datainrdy = 0;}  
    //33
    sfg dataset33In{datain = 81; datainrdy = 1;}
    sfg dataset33Out{$display("For dataset33, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 179, giving: ", dataout == 179);
                    datain = 0; datainrdy = 0;}
    sfg dataset33Wait{datain = 0; datainrdy = 0;}  
    //34
    sfg dataset34In{datain = 100; datainrdy = 1;}
    sfg dataset34Out{$display("For dataset34, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 182, giving: ", dataout == 182);
                    datain = 0; datainrdy = 0;}
    sfg dataset34Wait{datain = 0; datainrdy = 0;}  
    //35
    sfg dataset35In{datain = 100; datainrdy = 1;}
    sfg dataset35Out{$display("For dataset35, one gets that: dataout = ", $dec, dataout, ", where the expected value was = 185, giving: ", dataout == 185);
                    datain = 0; datainrdy = 0;}
    sfg dataset35Wait{datain = 0; datainrdy = 0;}  
}
 
 
 
fsm testBench(testUCP){
    initial sD1S;
    state sStall, sD1E, sD1W,
          sD2S, sD2E, sD2W,
          sD3S, sD3E, sD3W,
          sD4S, sD4E, sD4W,
          sD5S, sD5E, sD5W,
          sD6S, sD6E, sD6W,
          sD7S, sD7E, sD7W,
          sD8S, sD8E, sD8W,
          sD9S, sD9E, sD9W,
          sD10S, sD10E, sD10W,
          sD11S, sD11E, sD11W,
          sD12S, sD12E, sD12W,
          sD13S, sD13E, sD13W,
          sD14S, sD14E, sD14W,
          sD15S, sD15E, sD15W,
          sD16S, sD16E, sD16W,
          sD17S, sD17E, sD17W,
          sD18S, sD18E, sD18W,
          sD19S, sD19E, sD19W,          
          sD20S, sD20E, sD20W,
          sD21S, sD21E, sD21W,
          sD22S, sD22E, sD22W,
          sD23S, sD23E, sD23W,
          sD24S, sD24E, sD24W,
          sD25S, sD25E, sD25W,
          sD26S, sD26E, sD26W,
          sD27S, sD27E, sD27W,
          sD28S, sD28E, sD28W,
          sD29S, sD29E, sD29W,
          sD30S, sD30E, sD30W,
          sD31S, sD31E, sD31W,
          sD32S, sD32E, sD32W,
          sD33S, sD33E, sD33W,
          sD34S, sD34E, sD34W,
          sD35S, sD35E, sD35W;
    //1
    @sD1S              (dataset1In)     -> sD1E;
    @sD1E              (dataset1Out)    -> sD1W;
    @sD1W              (dataset1Wait)   -> sD2S;
    //2
    @sD2S              (dataset2In)     -> sD2E;
    @sD2E              (dataset2Out)    -> sD2W;
    @sD2W              (dataset2Wait)   -> sD3S;
    //3
    @sD3S              (dataset3In)     -> sD3E;
    @sD3E              (dataset3Out)    -> sD3W;
    @sD3W              (dataset3Wait)   -> sD4S;
    //4
    @sD4S              (dataset4In)     -> sD4E;
    @sD4E              (dataset4Out)    -> sD4W;
    @sD4W              (dataset4Wait)   -> sD5S;
    //5
    @sD5S              (dataset5In)     -> sD5E;
    @sD5E              (dataset5Out)    -> sD5W;
    @sD5W              (dataset5Wait)   -> sD6S;
    //6
    @sD6S              (dataset6In)     -> sD6E;
    @sD6E              (dataset6Out)    -> sD6W;
    @sD6W              (dataset6Wait)   -> sD7S;
    //7
    @sD7S              (dataset7In)     -> sD7E;
    @sD7E              (dataset7Out)    -> sD7W;
    @sD7W              (dataset7Wait)   -> sD8S;
    //8
    @sD8S              (dataset8In)     -> sD8E;
    @sD8E              (dataset8Out)    -> sD8W;
    @sD8W              (dataset8Wait)   -> sD9S;
    //9
    @sD9S              (dataset9In)     -> sD9E;
    @sD9E              (dataset9Out)    -> sD9W;
    @sD9W              (dataset9Wait)   -> sD10S;
    //10
    @sD10S              (dataset10In)     -> sD10E;
    @sD10E              (dataset10Out)    -> sD10W;
    @sD10W              (dataset10Wait)   -> sD11S;
    //11
    @sD11S              (dataset11In)     -> sD11E;
    @sD11E              (dataset11Out)    -> sD11W;
    @sD11W              (dataset11Wait)   -> sD12S;
    //12
    @sD12S              (dataset12In)     -> sD12E;
    @sD12E              (dataset12Out)    -> sD12W;
    @sD12W              (dataset12Wait)   -> sD13S;
    //13
    @sD13S              (dataset13In)     -> sD13E;
    @sD13E              (dataset13Out)    -> sD13W;
    @sD13W              (dataset13Wait)   -> sD14S;
    //14
    @sD14S              (dataset14In)     -> sD14E;
    @sD14E              (dataset14Out)    -> sD14W;
    @sD14W              (dataset14Wait)   -> sD15S;
    //15
    @sD15S              (dataset15In)     -> sD15E;
    @sD15E              (dataset15Out)    -> sD15W;
    @sD15W              (dataset15Wait)   -> sD16S;
    //16
    @sD16S              (dataset16In)     -> sD16E;
    @sD16E              (dataset16Out)    -> sD16W;
    @sD16W              (dataset16Wait)   -> sD17S;
    //17
    @sD17S              (dataset17In)     -> sD17E;
    @sD17E              (dataset17Out)    -> sD17W;
    @sD17W              (dataset17Wait)   -> sD18S;
    //18
    @sD18S              (dataset18In)     -> sD18E;
    @sD18E              (dataset18Out)    -> sD18W;
    @sD18W              (dataset18Wait)   -> sD19S;
    //19
    @sD19S              (dataset19In)     -> sD19E;
    @sD19E              (dataset19Out)    -> sD19W;
    @sD19W              (dataset19Wait)   -> sD20S;
    //20
    @sD20S              (dataset20In)     -> sD20E;
    @sD20E              (dataset20Out)    -> sD20W;
    @sD20W              (dataset20Wait)   -> sD21S;   
    //21
    @sD21S              (dataset21In)     -> sD21E;
    @sD21E              (dataset21Out)    -> sD21W;
    @sD21W              (dataset21Wait)   -> sD22S;   
    //22
    @sD22S              (dataset22In)     -> sD22E;
    @sD22E              (dataset22Out)    -> sD22W;
    @sD22W              (dataset22Wait)   -> sD23S;   
    //23
    @sD23S              (dataset23In)     -> sD23E;
    @sD23E              (dataset23Out)    -> sD23W;
    @sD23W              (dataset23Wait)   -> sD24S;   
    //24
    @sD24S              (dataset24In)     -> sD24E;
    @sD24E              (dataset24Out)    -> sD24W;
    @sD24W              (dataset24Wait)   -> sD25S;   
    //25
    @sD25S              (dataset25In)     -> sD25E;
    @sD25E              (dataset25Out)    -> sD25W;
    @sD25W              (dataset25Wait)   -> sD26S;   
    //26
    @sD26S              (dataset26In)     -> sD26E;
    @sD26E              (dataset26Out)    -> sD26W;
    @sD26W              (dataset26Wait)   -> sD27S;   
    //27
    @sD27S              (dataset27In)     -> sD27E;
    @sD27E              (dataset27Out)    -> sD27W;
    @sD27W              (dataset27Wait)   -> sD28S;   
    //28
    @sD28S              (dataset28In)     -> sD28E;
    @sD28E              (dataset28Out)    -> sD28W;
    @sD28W              (dataset28Wait)   -> sD29S;   
    //29
    @sD29S              (dataset29In)     -> sD29E;
    @sD29E              (dataset29Out)    -> sD29W;
    @sD29W              (dataset29Wait)   -> sD30S;   
    //30
    @sD30S              (dataset30In)     -> sD30E;
    @sD30E              (dataset30Out)    -> sD30W;
    @sD30W              (dataset30Wait)   -> sD31S;   
    //31
    @sD31S              (dataset31In)     -> sD31E;
    @sD31E              (dataset31Out)    -> sD31W;
    @sD31W              (dataset31Wait)   -> sD32S;   
    //32
    @sD32S              (dataset32In)     -> sD32E;
    @sD32E              (dataset32Out)    -> sD32W;
    @sD32W              (dataset32Wait)   -> sD33S;   
    //33
    @sD33S              (dataset33In)     -> sD33E;
    @sD33E              (dataset33Out)    -> sD33W;
    @sD33W              (dataset33Wait)   -> sD34S;
    //34
    @sD34S              (dataset34In)     -> sD34E;
    @sD34E              (dataset34Out)    -> sD34W;
    @sD34W              (dataset34Wait)   -> sD35S;   
    //35
    @sD35S              (dataset35In)     -> sD35E;
    @sD35E              (dataset35Out)    -> sD35W;
    @sD35W              (dataset35Wait)   -> sStall;      

    //Stall
    @sStall            (dataset1Wait)   -> sStall;
}

system testingSystem{
    mwiFilter(datain, datainrdy, dataout, dataoutrdy);
    testUCP(datain, datainrdy, dataout, dataoutrdy);
}
