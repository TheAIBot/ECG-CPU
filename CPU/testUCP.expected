ram: set size 32
ram: set wl 32
For dataset1, one gets that: dataout = 0, where the expected value was = 0, giving: 1
For dataset2, one gets that: dataout = 0, where the expected value was = 0, giving: 1
For dataset3, one gets that: dataout = 0, where the expected value was = 0, giving: 1
For dataset4, one gets that: dataout = 0, where the expected value was = 0, giving: 1
For dataset5, one gets that: dataout = 0, where the expected value was = 0, giving: 1
For dataset6, one gets that: dataout = 0, where the expected value was = 0, giving: 1
For dataset7, one gets that: dataout = 0, where the expected value was = 0, giving: 1
For dataset8, one gets that: dataout = 0, where the expected value was = 0 giving: 1
For dataset9, one gets that: dataout = 1, where the expected value was = 1, giving: 1
For dataset10, one gets that: dataout = 2, where the expected value was = 2, giving: 1
For dataset11, one gets that: dataout = 4, where the expected value was = 4, giving: 1
For dataset12, one gets that: dataout = 5, where the expected value was = 5, giving: 1
For dataset13, one gets that: dataout = 7, where the expected value was = 7, giving: 1
For dataset14, one gets that: dataout = 8, where the expected value was = 8, giving: 1
For dataset15, one gets that: dataout = 10, where the expected value was = 10, giving: 1
For dataset16, one gets that: dataout = 12, where the expected value was = 12, giving: 1
For dataset17, one gets that: dataout = 12, where the expected value was = 12, giving: 1
For dataset18, one gets that: dataout = 12, where the expected value was = 12, giving: 1
For dataset19, one gets that: dataout = 13, where the expected value was = 13, giving: 1
For dataset20, one gets that: dataout = 18, where the expected value was = 18, giving: 1
For dataset21, one gets that: dataout = 30, where the expected value was = 30, giving: 1
For dataset22, one gets that: dataout = 55, where the expected value was = 55, giving: 1
For dataset23, one gets that: dataout = 89, where the expected value was = 89, giving: 1
For dataset24, one gets that: dataout = 123, where the expected value was = 123, giving: 1
For dataset25, one gets that: dataout = 149, where the expected value was = 149, giving: 1
For dataset26, one gets that: dataout = 164, where the expected value was = 164, giving: 1
For dataset27, one gets that: dataout = 170, where the expected value was = 170, giving: 1
For dataset28, one gets that: dataout = 171, where the expected value was = 171, giving: 1
For dataset29, one gets that: dataout = 171, where the expected value was = 171, giving: 1
For dataset30, one gets that: dataout = 172, where the expected value was = 172, giving: 1
For dataset31, one gets that: dataout = 174, where the expected value was = 174, giving: 1
For dataset32, one gets that: dataout = 176, where the expected value was = 176, giving: 1
For dataset33, one gets that: dataout = 179, where the expected value was = 179, giving: 1
For dataset34, one gets that: dataout = 182, where the expected value was = 182, giving: 1
For dataset35, one gets that: dataout = 185, where the expected value was = 185, giving: 1
